From 84d8a0ca5b521b9d87679ffebe420fe69869961d Mon Sep 17 00:00:00 2001
From: Kinglong Mee <kinglongmee@gmail.com>
Date: Mon, 8 Jul 2019 21:13:28 +0800
Subject: [PATCH 270/276] cluster/ec: inherit healing from lock when it has
 info

If lock has info, fop should inherit healing mask from it.
Otherwise, fop cannot inherit right healing when changed_flags is zero.

Upstream-patch: https://review.gluster.org/c/glusterfs/+/23010
Change-Id: Ife80c9169d2c555024347a20300b0583f7e8a87f
fixes: bz#1732792
Signed-off-by: Kinglong Mee <mijinlong@horiscale.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/177974
Tested-by: RHGS Build Bot <nigelb@redhat.com>
Reviewed-by: Sunil Kumar Heggodu Gopala Acharya <sheggodu@redhat.com>
---
 xlators/cluster/ec/src/ec-common.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/xlators/cluster/ec/src/ec-common.c b/xlators/cluster/ec/src/ec-common.c
index e2e582f..db1ff5b 100644
--- a/xlators/cluster/ec/src/ec-common.c
+++ b/xlators/cluster/ec/src/ec-common.c
@@ -1412,11 +1412,12 @@ ec_get_size_version(ec_lock_link_t *link)
     set_dirty = ec_set_dirty_flag(link, ctx, dirty);
 
     /* If ec metadata has already been retrieved, do not try again. */
-    if (ctx->have_info && (!set_dirty)) {
+    if (ctx->have_info) {
         if (ec_is_data_fop(fop->id)) {
             fop->healing |= lock->healing;
         }
-        goto unlock;
+        if (!set_dirty)
+            goto unlock;
     }
 
     /* Determine if there's something we need to retrieve for the current
-- 
1.8.3.1

